### ISSUE TITLE
Sharing entire screen on whereby with Firefox 78.0.1 and Chromium 83.0.4103.116 shows black screen

### ISSUE THREAD

**Issue by devurandom** [emersion/xdg-desktop-portal-wlr#41]
# Problem

Sharing the entire screen in Firefox or Chromium shows only a black screen. The behaviour with regards to updates / moving the mouse is odd.

# Versions used

* Firefox 78.0.1 with pipewire-0.3 patch
* Chromium 83.0.4103.116
* pipewire 0.3.6
* xdg-desktop-portal 1.7.2
* xdg-desktop-portal-wlr bc89778fdc2f20323a9201bb5d17be639c181268

# Steps to prepare the desktop

1. Start Sway
2. Kill `xdg-desktop-portal` and any implementation that might be running (e.g. `xdg-desktop-portal-gtk`)
   - Started by systemd, thus has no access to the `XDG_CURRENT_DESKTOP` environment variable set in the script starting Sway
3. Start `/usr/libexec/xdg-desktop-portal -vr` in a terminal started
   - Please find the output below

# Window setup in Sway

The Firefox Whereby tab occupies the left half of the screen. On the right half I have Chromium, Alacritty and a few more Firefox windows stacked.

# Steps for each browser

1. Open https://whereby.com/<myroom>
2. Click "Share" in the mid-bottom of the page

# Behaviour when clicking "Share"

## Firefox

No windows are displayed in the list.

Firefox generally displays no preview. Also not on X11.

## Chromium

I can select "entire screen". The preview is black with just the mouse cursor visible and moving.

The Chromium window is selectable. The preview shows the content of the actual window, i.e. it is "live" and reacts to changes in the page, including the share window selection "window" itself. The mouse cursor is not visible.

No other windows are listed.

# Behaviour when streaming

The behaviour is generally the same with both browsers, modulo the fact that Firefox cannot share windows.

## Previews vs. actual stream

The preview displayed by both browsers is the same as what the other browser connected to the video call shows.

## Share entire screen

The video feed is black with just the mouse cursor visible. The mouse cursor is stuck while I move it over Firefox or Alacritty windows. Movement is only shown when I move the cursor over the Chromium window. This is independently of whether Chromium has my Whereby room open or not, i.e. I can reproduce it by moving the mouse cursor over a freshly started Chromium showing only one "new tab" page.

## Share window

Since this does not work in Firefox, this section describes the behaviour when sharing a window from Chromium!

The video feed shows the actual content of the window. The feed is live and reacts to changes in window content, like switching tabs or watching a video on YouTube. The mouse cursor is visible and moving. When switching to a different window, so that Chromium is no longer visible, the mouse cursor vanishes and after up to a few seconds (but sometimes immediately) the update rate gets reduced to <1fps when a YouTube video is playing, i.e. in Whereby I see only snapshots of the video every second or less.

# Output of `/usr/libexec/xdg-desktop-portal -vr`

```
XDP: load portals from /usr/share/xdg-desktop-portal/portals
XDP: loading /usr/share/xdg-desktop-portal/portals/gtk.portal
XDP: portal implementation for gnome
XDP: portal implementation supports org.freedesktop.impl.portal.FileChooser
XDP: portal implementation supports org.freedesktop.impl.portal.AppChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Print
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.Notification
XDP: portal implementation supports org.freedesktop.impl.portal.Inhibit
XDP: portal implementation supports org.freedesktop.impl.portal.Access
XDP: portal implementation supports org.freedesktop.impl.portal.Account
XDP: portal implementation supports org.freedesktop.impl.portal.Email
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: portal implementation supports org.freedesktop.impl.portal.RemoteDesktop
XDP: portal implementation supports org.freedesktop.impl.portal.Lockdown
XDP: portal implementation supports org.freedesktop.impl.portal.Background
XDP: portal implementation supports org.freedesktop.impl.portal.Settings
XDP: portal implementation supports org.freedesktop.impl.portal.Wallpaper
XDP: loading /usr/share/xdg-desktop-portal/portals/wlr.portal
XDP: portal implementation for sway
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: loading /usr/share/xdg-desktop-portal/portals/kde.portal
XDP: portal implementation for KDE
XDP: portal implementation supports org.freedesktop.impl.portal.Access
XDP: portal implementation supports org.freedesktop.impl.portal.Account
XDP: portal implementation supports org.freedesktop.impl.portal.AppChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Background
XDP: portal implementation supports org.freedesktop.impl.portal.Email
XDP: portal implementation supports org.freedesktop.impl.portal.FileChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Inhibit
XDP: portal implementation supports org.freedesktop.impl.portal.Notification
XDP: portal implementation supports org.freedesktop.impl.portal.Print
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.RemoteDesktop
XDP: portal implementation supports org.freedesktop.impl.portal.Settings
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Lockdown
XDP: providing portal org.freedesktop.portal.MemoryMonitor
XDP: providing portal org.freedesktop.portal.NetworkMonitor
XDP: providing portal org.freedesktop.portal.ProxyResolver
XDP: providing portal org.freedesktop.portal.Trash
XDP: providing portal org.freedesktop.portal.GameMode
XDP: Using gtk.portal for org.freedesktop.impl.portal.Settings
XDP: Using kde.portal for org.freedesktop.impl.portal.Settings

** (/usr/libexec/xdg-desktop-portal:1671141): WARNING **: 09:11:25.268: Failed to create settings proxy: Error calling StartServiceByName for org.freedesktop.impl.portal.desktop.kde: Timeout was reached
XDP: providing portal org.freedesktop.portal.Settings
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.FileChooser
XDP: providing portal org.freedesktop.portal.FileChooser
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.AppChooser
XDP: providing portal org.freedesktop.portal.OpenURI
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Print
XDP: providing portal org.freedesktop.portal.Print
XDP: Using wlr.portal for org.freedesktop.impl.portal.Screenshot in sway
XDP: providing portal org.freedesktop.portal.Screenshot
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Notification
XDP: providing portal org.freedesktop.portal.Notification
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Inhibit
XDP: providing portal org.freedesktop.portal.Inhibit
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Access
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Background
XDP: providing portal org.freedesktop.portal.Device
XDP: providing portal org.freedesktop.portal.Location
XDP: providing portal org.freedesktop.portal.Camera
XDP: Starting background app monitor
XDP: providing portal org.freedesktop.portal.Background
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Wallpaper
XDP: providing portal org.freedesktop.portal.Wallpaper
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Account
XDP: using org.freedesktop.impl.portal.Account at org.freedesktop.impl.portal.desktop.gtk

XDP: providing portal org.freedesktop.portal.Account
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Email
XDP: providing portal org.freedesktop.portal.Email
XDP: Using wlr.portal for org.freedesktop.impl.portal.ScreenCast in sway
XDP: providing portal org.freedesktop.portal.ScreenCast
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.RemoteDesktop
XDP: providing portal org.freedesktop.portal.RemoteDesktop
XDP: org.freedesktop.portal.Desktop acquired

** (/usr/libexec/xdg-desktop-portal:1671141): WARNING **: 09:11:55.313: Failed to get application states: GDBus.Error:org.freedesktop.portal.Error.Failed: Could not get window list
```

**Comment by danshick** [comment]
This sounds like you aren't setting XDG_SESSION_TYPE=wayland before starting your browser.

If that doesn't fix it, can you start xdpw as described [here](https://github.com/emersion/xdg-desktop-portal-wlr/wiki/FAQ#but-i-need-to-use-a-cli-flag-why-cant-i-just-run-xdpw-directly), and use the `-l DEBUG` flag to produce logs from xdpw itself?

It would also be helpful if you could test with the [gstreamer python script](https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#xdpgstreamer-python-script) to rule out browser issues.

**Comment by devurandom** [comment]
> This sounds like you aren't setting XDG_SESSION_TYPE=wayland before starting your browser.

Sorry, I forgot to mention that I start Sway with these variables set:

```
# Session
export XDG_SESSION_TYPE=wayland
export XDG_SESSION_DESKTOP=sway
export XDG_CURRENT_DESKTOP=sway

# Qt
export QT_QPA_PLATFORM="wayland-egl"
export QT_WAYLAND_FORCE_DPI="physical"
export QT_WAYLAND_DISABLEWINDOWDECORATION=1

# Gnome
export CLUTTER_BACKEND="wayland"

# Enlightenment
export ECORE_EVAS_ENGINE="wayland-egl"
export ELM_ENGINE="wayland_egl"

# SDL
export SDL_VIDEODRIVER="wayland"

# Firefox
export MOZ_ENABLE_WAYLAND=1
export MOZ_DBUS_REMOTE=1

# Java
export _JAVA_AWT_WM_NONREPARENTING=1

# AT-SPI
export NO_AT_BRIDGE=1
```

I can confirm that they are set inside the session:

```
❯ env | grep XDG
XDG_CONFIG_DIRS=/etc/xdg
XDG_CURRENT_DESKTOP=sway
XDG_DATA_DIRS=/home/dschridde/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share
XDG_RUNTIME_DIR=/run/user/1000
XDG_SEAT=seat0
XDG_SEAT_PATH=/org/freedesktop/DisplayManager/Seat0
XDG_SESSION_CLASS=user
XDG_SESSION_DESKTOP=sway
XDG_SESSION_ID=1
XDG_SESSION_PATH=/org/freedesktop/DisplayManager/Session1
XDG_SESSION_TYPE=wayland
XDG_VTNR=2
```

Starting Firefox from the terminal that printed this makes no difference.

> If that doesn't fix it, can you start xdpw as described [here](https://github.com/emersion/xdg-desktop-portal-wlr/wiki/FAQ#but-i-need-to-use-a-cli-flag-why-cant-i-just-run-xdpw-directly), and use the `-l DEBUG` flag to produce logs from xdpw itself?

I did that (without restarting Firefox), this is the output:

```
❯ /usr/libexec/xdg-desktop-portal -vr
XDP: load portals from /usr/share/xdg-desktop-portal/portals
XDP: loading /usr/share/xdg-desktop-portal/portals/gtk.portal
XDP: portal implementation for gnome
XDP: portal implementation supports org.freedesktop.impl.portal.FileChooser
XDP: portal implementation supports org.freedesktop.impl.portal.AppChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Print
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.Notification
XDP: portal implementation supports org.freedesktop.impl.portal.Inhibit
XDP: portal implementation supports org.freedesktop.impl.portal.Access
XDP: portal implementation supports org.freedesktop.impl.portal.Account
XDP: portal implementation supports org.freedesktop.impl.portal.Email
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: portal implementation supports org.freedesktop.impl.portal.RemoteDesktop
XDP: portal implementation supports org.freedesktop.impl.portal.Lockdown
XDP: portal implementation supports org.freedesktop.impl.portal.Background
XDP: portal implementation supports org.freedesktop.impl.portal.Settings
XDP: portal implementation supports org.freedesktop.impl.portal.Wallpaper
XDP: loading /usr/share/xdg-desktop-portal/portals/wlr.portal
XDP: portal implementation for sway
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: loading /usr/share/xdg-desktop-portal/portals/kde.portal
XDP: portal implementation for KDE
XDP: portal implementation supports org.freedesktop.impl.portal.Access
XDP: portal implementation supports org.freedesktop.impl.portal.Account
XDP: portal implementation supports org.freedesktop.impl.portal.AppChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Background
XDP: portal implementation supports org.freedesktop.impl.portal.Email
XDP: portal implementation supports org.freedesktop.impl.portal.FileChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Inhibit
XDP: portal implementation supports org.freedesktop.impl.portal.Notification
XDP: portal implementation supports org.freedesktop.impl.portal.Print
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.RemoteDesktop
XDP: portal implementation supports org.freedesktop.impl.portal.Settings
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Lockdown
XDP: providing portal org.freedesktop.portal.MemoryMonitor
XDP: providing portal org.freedesktop.portal.NetworkMonitor
XDP: providing portal org.freedesktop.portal.ProxyResolver
XDP: providing portal org.freedesktop.portal.Trash
XDP: providing portal org.freedesktop.portal.GameMode
XDP: Using gtk.portal for org.freedesktop.impl.portal.Settings
XDP: Using kde.portal for org.freedesktop.impl.portal.Settings

** (/usr/libexec/xdg-desktop-portal:5043): WARNING **: 14:34:46.318: Failed to create settings proxy: Error calling StartServiceByName for org.freedesktop.impl.portal.desktop.kde: Timeout was reached
XDP: providing portal org.freedesktop.portal.Settings
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.FileChooser
XDP: providing portal org.freedesktop.portal.FileChooser
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.AppChooser
XDP: providing portal org.freedesktop.portal.OpenURI
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Print
XDP: providing portal org.freedesktop.portal.Print
XDP: Using wlr.portal for org.freedesktop.impl.portal.Screenshot in sway
XDP: providing portal org.freedesktop.portal.Screenshot
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Notification
XDP: providing portal org.freedesktop.portal.Notification
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Inhibit
XDP: providing portal org.freedesktop.portal.Inhibit
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Access
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Background
XDP: providing portal org.freedesktop.portal.Device
XDP: providing portal org.freedesktop.portal.Location
XDP: providing portal org.freedesktop.portal.Camera
XDP: Starting background app monitor
XDP: providing portal org.freedesktop.portal.Background
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Wallpaper
XDP: providing portal org.freedesktop.portal.Wallpaper
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Account
XDP: using org.freedesktop.impl.portal.Account at org.freedesktop.impl.portal.desktop.gtk

XDP: providing portal org.freedesktop.portal.Account
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Email
XDP: providing portal org.freedesktop.portal.Email
XDP: Using wlr.portal for org.freedesktop.impl.portal.ScreenCast in sway
XDP: providing portal org.freedesktop.portal.ScreenCast
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.RemoteDesktop
XDP: providing portal org.freedesktop.portal.RemoteDesktop
XDP: org.freedesktop.portal.Desktop acquired

** (/usr/libexec/xdg-desktop-portal:5043): WARNING **: 14:35:16.326: Failed to get application states: GDBus.Error:org.freedesktop.portal.Error.Failed: Could not get window list
```

```
❯ /usr/libexec/xdg-desktop-portal-wlr -l DEBUG
2020/07/11 14:34:35 [DEBUG] - dbus: connected
2020/07/11 14:34:35 [DEBUG] - wlroots: wl_display connected
2020/07/11 14:34:35 [DEBUG] - pipewire: pw_loop created
2020/07/11 14:34:35 [DEBUG] - pipewire: establishing connection to core
2020/07/11 14:34:35 [DEBUG] - wayland: registry listeners run
2020/07/11 14:34:35 [DEBUG] - wayland: xdg output listeners run
```

> It would also be helpful if you could test with the [gstreamer python script](https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#xdpgstreamer-python-script) to rule out browser issues.

The script outputs the following and shows the content of my entire screen:

```
❯ python3.7 ./bin/xdp-screen-cast.py
./bin/xdp-screen-cast.py:13: PyGIDeprecationWarning: Since version 3.11, calling threads_init is no longer needed. See: https://wiki.gnome.org/PyGObject/Threading
  GObject.threads_init()
session /org/freedesktop/portal/desktop/session/1_140/u1 created
sources selected
streams:
stream 4294967295
```

```
❯ /usr/libexec/xdg-desktop-portal -vr
XDP: load portals from /usr/share/xdg-desktop-portal/portals
XDP: loading /usr/share/xdg-desktop-portal/portals/gtk.portal
XDP: portal implementation for gnome
XDP: portal implementation supports org.freedesktop.impl.portal.FileChooser
XDP: portal implementation supports org.freedesktop.impl.portal.AppChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Print
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.Notification
XDP: portal implementation supports org.freedesktop.impl.portal.Inhibit
XDP: portal implementation supports org.freedesktop.impl.portal.Access
XDP: portal implementation supports org.freedesktop.impl.portal.Account
XDP: portal implementation supports org.freedesktop.impl.portal.Email
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: portal implementation supports org.freedesktop.impl.portal.RemoteDesktop
XDP: portal implementation supports org.freedesktop.impl.portal.Lockdown
XDP: portal implementation supports org.freedesktop.impl.portal.Background
XDP: portal implementation supports org.freedesktop.impl.portal.Settings
XDP: portal implementation supports org.freedesktop.impl.portal.Wallpaper
XDP: loading /usr/share/xdg-desktop-portal/portals/wlr.portal
XDP: portal implementation for sway
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: loading /usr/share/xdg-desktop-portal/portals/kde.portal
XDP: portal implementation for KDE
XDP: portal implementation supports org.freedesktop.impl.portal.Access
XDP: portal implementation supports org.freedesktop.impl.portal.Account
XDP: portal implementation supports org.freedesktop.impl.portal.AppChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Background
XDP: portal implementation supports org.freedesktop.impl.portal.Email
XDP: portal implementation supports org.freedesktop.impl.portal.FileChooser
XDP: portal implementation supports org.freedesktop.impl.portal.Inhibit
XDP: portal implementation supports org.freedesktop.impl.portal.Notification
XDP: portal implementation supports org.freedesktop.impl.portal.Print
XDP: portal implementation supports org.freedesktop.impl.portal.ScreenCast
XDP: portal implementation supports org.freedesktop.impl.portal.Screenshot
XDP: portal implementation supports org.freedesktop.impl.portal.RemoteDesktop
XDP: portal implementation supports org.freedesktop.impl.portal.Settings
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Lockdown
XDP: providing portal org.freedesktop.portal.MemoryMonitor
XDP: providing portal org.freedesktop.portal.NetworkMonitor
XDP: providing portal org.freedesktop.portal.ProxyResolver
XDP: providing portal org.freedesktop.portal.Trash
XDP: providing portal org.freedesktop.portal.GameMode
XDP: Using gtk.portal for org.freedesktop.impl.portal.Settings
XDP: Using kde.portal for org.freedesktop.impl.portal.Settings

** (/usr/libexec/xdg-desktop-portal:6472): WARNING **: 14:39:52.624: Failed to create settings proxy: Error calling StartServiceByName for org.freedesktop.impl.portal.desktop.kde: Timeout was reached
XDP: providing portal org.freedesktop.portal.Settings
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.FileChooser
XDP: providing portal org.freedesktop.portal.FileChooser
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.AppChooser
XDP: providing portal org.freedesktop.portal.OpenURI
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Print
XDP: providing portal org.freedesktop.portal.Print
XDP: Using wlr.portal for org.freedesktop.impl.portal.Screenshot in sway
XDP: providing portal org.freedesktop.portal.Screenshot
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Notification
XDP: providing portal org.freedesktop.portal.Notification
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Inhibit
XDP: providing portal org.freedesktop.portal.Inhibit
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Access
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Background
XDP: providing portal org.freedesktop.portal.Device
XDP: providing portal org.freedesktop.portal.Location
XDP: providing portal org.freedesktop.portal.Camera
XDP: Starting background app monitor
XDP: providing portal org.freedesktop.portal.Background
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Wallpaper
XDP: providing portal org.freedesktop.portal.Wallpaper
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Account
XDP: using org.freedesktop.impl.portal.Account at org.freedesktop.impl.portal.desktop.gtk

XDP: providing portal org.freedesktop.portal.Account
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.Email
XDP: providing portal org.freedesktop.portal.Email
XDP: Using wlr.portal for org.freedesktop.impl.portal.ScreenCast in sway
XDP: providing portal org.freedesktop.portal.ScreenCast
XDP: Falling back to gtk.portal for org.freedesktop.impl.portal.RemoteDesktop
XDP: providing portal org.freedesktop.portal.RemoteDesktop
XDP: org.freedesktop.portal.Desktop acquired

** (/usr/libexec/xdg-desktop-portal:6472): WARNING **: 14:40:22.639: Failed to get application states: GDBus.Error:org.freedesktop.portal.Error.Failed: Could not get window list
XDP: screen cast session owned by ':1.147' created
XDP: screen cast session owned by ':1.147' started
[W][000000802.315173][core.c:191 destroy_proxy()] core 0x7f374c027f40: leaked proxy 0x7f374c03b050 id:2
```

```
❯ /usr/libexec/xdg-desktop-portal-wlr -l DEBUG
2020/07/11 14:39:24 [DEBUG] - dbus: connected
2020/07/11 14:39:24 [DEBUG] - wlroots: wl_display connected
2020/07/11 14:39:24 [DEBUG] - pipewire: pw_loop created
2020/07/11 14:39:24 [DEBUG] - pipewire: establishing connection to core
2020/07/11 14:39:24 [DEBUG] - wayland: registry listeners run
2020/07/11 14:39:24 [DEBUG] - wayland: xdg output listeners run
2020/07/11 14:40:59 [INFO] - dbus: create session method invoked
2020/07/11 14:40:59 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_147/u1
2020/07/11 14:40:59 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_147/u1
2020/07/11 14:40:59 [INFO] - dbus: app_id:
2020/07/11 14:40:59 [INFO] - dbus: select sources method invoked
2020/07/11 14:40:59 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_147/u2
2020/07/11 14:40:59 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_147/u1
2020/07/11 14:40:59 [INFO] - dbus: app_id:
2020/07/11 14:40:59 [INFO] - dbus: option types:3
2020/07/11 14:40:59 [INFO] - dbus: option multiple: 0
2020/07/11 14:40:59 [DEBUG] - dbus: select sources: found matching session /org/freedesktop/portal/desktop/session/1_147/u1
2020/07/11 14:40:59 [INFO] - wlroots: capturable output: Goldstar Company Ltd model: LG ULTRAWIDE: id: 30 name: DP-2
2020/07/11 14:40:59 [INFO] - xdpw: screencast instance 0x55aa089ee950 has 1 references
2020/07/11 14:40:59 [INFO] - xdpw: 1 active screencast instances
2020/07/11 14:40:59 [INFO] - wlroots: output: DP-2
2020/07/11 14:40:59 [INFO] - dbus: start method invoked
2020/07/11 14:40:59 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_147/u3
2020/07/11 14:40:59 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_147/u1
2020/07/11 14:40:59 [INFO] - dbus: app_id:
2020/07/11 14:40:59 [INFO] - dbus: parent_window:
2020/07/11 14:40:59 [DEBUG] - dbus: start: found matching session /org/freedesktop/portal/desktop/session/1_147/u1
2020/07/11 14:40:59 [DEBUG] - pipewire: registered event 0x55aa089ef4f0
2020/07/11 14:40:59 [INFO] - pipewire: stream state changed to "connecting"
2020/07/11 14:40:59 [INFO] - pipewire: node id is -1
2020/07/11 14:40:59 [INFO] - pipewire: stream state changed to "paused"
2020/07/11 14:40:59 [INFO] - pipewire: node id is 84
2020/07/11 14:40:59 [INFO] - pipewire: stream state changed to "streaming"
2020/07/11 14:40:59 [INFO] - pipewire: node id is 84
```

**Comment by devurandom** [comment]
Do I read the difference in `xdg-desktop-portal-wlr`'s output between firefox and script No. 19 correctly in that Firefox never tries to contact the WLR portal?

**Comment by devurandom** [comment]
> Do I read the difference in `xdg-desktop-portal-wlr`'s output between firefox and script No. 19 correctly in that Firefox never tries to contact the WLR portal?

I think I might have found the culprit. The patches were included into Gentoo's package, but the build option was by default disabled. I will report back whether force-enabling the build option solves the issue, once I rebuilt Firefox.

**Comment by danshick** [comment]
I'm not a Gentoo user anymore, but I worked with the user who has helped in adding Gentoo compatibility. They are prometheanfire on #sway on IRC. If you're having trouble with the Gentoo patches, they might be willing to help.

Also yes, I think you're right. The python script working definitely makes this an FF problem.

**Comment by devurandom** [comment]
At first Firefox would segfault when trying to share the screen. I now rebuilt media-video/pipewire-0.3.6, sys-apps/xdg-desktop-portal-1.7.2, gui-libs/xdg-desktop-portal-wlr-9999, gui-wm/sway-1.4-r2 and www-client/firefox-78.0.2 from the Gentoo repository. Now I can screenshare in Firefox in Sway on Gentoo. :)

@prometheanfire already updated the compatibility list: https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility/_compare/54d32b7d4af110299a9feb711ba3e6df31ebdcc1...970def9ea0dc0e77aba5f56c41b5dbd971c3e700

I just added the versions I tested with to that list.

Thank you for your help, @danshick!
